File: Lab2/CSharp/MergeSortTests.cs
using System.Linq;
using FluentAssertions;
using NUnit.Framework;

namespace CSharp {
    [TestFixture]
    public class MergeSortTests {
        private static (int[] source, int pieces, int[][] expected)[] _splitIntoSource = TestSources.SplitIntoSource;
        
        private static (int[] left, int[] right, int[] expected)[] _mergeSorce = TestSources.MergeSorce;
        
        private static int[][] _sortSorce= TestSources.SortSorce;
        
        
        [TestCaseSource(nameof(_splitIntoSource))]
        public void SplitIntoTest((int[] source, int pieces, int[][] expected) args) {
            var result = args.source.SplitInto(args.pieces);
            
            result.Should().BeEquivalentTo(args.expected);
        }

        [TestCaseSource(nameof(_mergeSorce))]
        public void MergeSorceTest((int[] left, int[] right, int[] expected) args) {
            var result = args.left.Merge(args.right);
            
            result.Should().BeEquivalentTo(args.expected);
        }

        [TestCaseSource(nameof(_sortSorce))]
        public void QuickSortTest(int[] array) {
            var result = array.MergeSort();
            var expected = array.OrderBy(x => x).ToArray();
            result.Should().BeEquivalentTo(expected);
        }
    }
}

File: Lab2/FSharp/MergeSortTests.fs
module MergeSortTests

open NUnit.Framework
open FsUnit

[<TestFixture>]
type Tests() =
    static member SortArraySource = TestSources.sortSource
    
    static member SortSeqSource = TestSources.sortSource
    
    static member MergeListSource = TestSources.mergeSource
    
    static member MergeSeqSource = TestSources.mergeSource
    
    [<TestCaseSource("SortArraySource")>]
    static member MergeSortArrayTest list =
        let array = List.toArray list         
        let result = MergeSort.mergeSort_v1List array

        let expected = Array.sort array
        result |> should equal expected

    [<TestCaseSource("SortSeqSource")>]
    static member MergeSortSeqTest list =
        let array = List.toArray list
        let result = MergeSort.mergeSort_v2Seq array |> Seq.toArray

        let expected = Array.sort array
        result |> should equal expected

    [<TestCaseSource("MergeListSource")>]
    static member MergeListTest (tuple : 'a list * 'a list * 'a list) =
        let (left, right, expected) = tuple 
        let actual = MergeSort.merge_v1List left right
        
        actual |> should equal expected

    [<TestCaseSource("MergeSeqSource")>]
    static member MergeSeqTest (tuple : 'a list * 'a list * 'a list) =
        let (left, right, expected) = tuple 
        let actual = MergeSort.merge_v2Seq left right
        
        actual |> should equal expected
    
        

